src/linters/spacing.js

```diff
@@ -0,0 +1,124 @@
+/* eslint-disable import/no-commonjs */
+
+const matchAll = require('string.prototype.matchall');
+const fs = require('fs');
+const path = require('path');
+const stylelint = require('stylelint');
+
+const spacing = fs.readFileSync(path.join(__dirname, '..', 'styles', 'spacing.css'), 'UTF-8');
+const cssVarRegex = /--[a-z0-9-]+/g;
+const validSpacingSettings = spacing.match(cssVarRegex);
+
+const cssValueRegex = /[0-9]+px/g;
+const valueKeys = spacing.match(cssValueRegex);
+
+const ruleName = 'mds/spacing';
+const url = 'https://mavenlink.github.io/design-system/master/#/Brand%20Identity/Spacing';
+
+function zip(obj, value, index) {
+  return { ...obj, [value]: validSpacingSettings[index] };
+}
+const fixmap = valueKeys.reduce(zip, {});
+
+const messages = stylelint.utils.ruleMessages(ruleName, {
+  rejected: `To promote design consistency: use MDS spacing variables for margin and padding.
+This rule can always be auto fixed with the --fix option.
+Ie. ${JSON.stringify(fixmap)} For more info, see: ${url}.
+If you are using a spacing value that is close to the MDS variable value please consider using the MDS value.`,
+});
+
+const properties = [
+  'margin',
+  'margin-top',
+  'margin-right',
+  'margin-bottom',
+  'margin-left',
+  'padding',
+  'padding-top',
+  'padding-right',
+  'padding-bottom',
+  'padding-left',
+];
+
+const fixes = [];
+
+module.exports = stylelint.createPlugin(ruleName, (primary, secondary, context) => {
+  return (root, result) => {
+    root.walkDecls((declaration) => {
+      const property = declaration.prop;
+
+      if (properties.includes(property)) {
+        const currentValue = declaration.value;
+
+        const invalidSpacingValue = Object.keys(fixmap).some((spacingVariable) => {
+          const pixelValueMatches = [...matchAll(currentValue, /\d+px/g)];
```

Comment:
I am not sure if we want to include yet another library as a dependency for this use case. It looks like `matchAll` is just providing a different return structure:

```
const regexp = /\d+px/g;
const str = 'blue 10px black 11px';

console.log([...str.matchAll(regexp)]); // Array [Array ["10px"], Array ["11px"]]
console.log(str.match(regexp)) // Array ["10px", "11px"]
```

I would prefer to use ES5 native methods instead of pulling in another library. Do we think we can refactor this to use `match`?

Reply:
The structure that `matchAll` returns is quite useful here because it includes additional information that we need. Specifically, that is the index of the match.

Reply:
Yes, we can use match but we will be nearly re-implementing `matchAll`.